src/Servers/Kestrel/Core/src/Internal/Http3/Helpers/VariableIntHelper.cs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3
+{
+    public static class VariableIntHelper
+    {
+        public static long GetVariableIntErrorIfNotAllowedValue(in ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
+        {
+            // 0x1F * N + 0x21 isn't allowed by spec, return -1 if that's the case.
+            // https://quicwg.org/base-drafts/draft-ietf-quic-http.html#frame-reserved
+            var longLength = GetVariableIntFromReadOnlySequence(buffer, out consumed, out examined);
+            if ((longLength - 0x21) % 0x1F == 0)
+            {
+                return -1;
+            }
+
+            return longLength;
+        }
+
+        public static long GetVariableIntFromReadOnlySequence(in ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
```

Comment:
Use `BinaryPrimitives` here.

Reply:
It's actually fairly awkward to use BinaryPrimitives here, I need to read the first byte to figure out the length, then read the rest of the integer (which is either 1 byte, 3 bytes, or 7 bytes).

Reply:
It will be more efficient to use `BinaryPrimitives` and re-read that first byte in the 4/8 byte case. I would benchmark if unsure.

https://github.com/aspnet/AspNetCore/blob/4a1d909c03a3b8e800239a21e5ad241ff8cb88f3/src/Servers/Kestrel/Core/src/Internal/Http3/Helpers/VariableIntHelper.cs#L61-L64

https://github.com/aspnet/AspNetCore/blob/4a1d909c03a3b8e800239a21e5ad241ff8cb88f3/src/Servers/Kestrel/Core/src/Internal/Http3/Helpers/VariableIntHelper.cs#L76-L83

Reply:
I trust it is faster 😄

Reply:
Profiling can come in another PR.